src/dev.rs

```diff
@@ -295,47 +295,47 @@ impl<F: FieldExt> MockProver<F> {
         let n = self.n as i32;
 
         // Check that all gates are satisfied for all rows.
-        let gate_errors =
-            self.cs
-                .gates
-                .iter()
-                .enumerate()
-                .flat_map(|(gate_index, (gate_name, gate))| {
-                    // We iterate from n..2n so we can just reduce to handle wrapping.
-                    (n..(2 * n)).filter_map(move |row| {
-                        fn load<'a, F: FieldExt, T: ColumnType>(
-                            n: i32,
-                            row: i32,
-                            queries: &'a [(Column<T>, Rotation)],
-                            cells: &'a [Vec<F>],
-                        ) -> impl Fn(usize) -> F + 'a {
-                            move |index| {
-                                let (column, at) = &queries[index];
-                                let resolved_row = (row + at.0) % n;
-                                cells[column.index()][resolved_row as usize]
-                            }
-                        }
-
-                        if gate.evaluate(
-                            &|scalar| scalar,
-                            &load(n, row, &self.cs.fixed_queries, &self.fixed),
-                            &load(n, row, &self.cs.advice_queries, &self.advice),
-                            &load(n, row, &self.cs.instance_queries, &self.instance),
-                            &|a, b| a + &b,
-                            &|a, b| a * &b,
-                            &|a, scalar| a * scalar,
-                        ) == F::zero()
-                        {
-                            None
-                        } else {
-                            Some(VerifyFailure::Gate {
-                                gate_index,
-                                gate_name,
-                                row: (row - n) as usize,
-                            })
+        let gate_errors = self
+            .cs
+            .gates
+            .iter()
+            .enumerate()
+            .flat_map(|(gate_index, gate)| {
+                // We iterate from n..2n so we can just reduce to handle wrapping.
+                (n..(2 * n)).filter_map(move |row| {
+                    fn load<'a, F: FieldExt, T: ColumnType>(
+                        n: i32,
+                        row: i32,
+                        queries: &'a [(Column<T>, Rotation)],
+                        cells: &'a [Vec<F>],
+                    ) -> impl Fn(usize) -> F + 'a {
+                        move |index| {
+                            let (column, at) = &queries[index];
+                            let resolved_row = (row + at.0) % n;
+                            cells[column.index()][resolved_row as usize]
                         }
-                    })
-                });
+                    }
+
+                    if gate.poly().evaluate(
```

Comment:
[In the halo2 book](https://zcash.github.io/halo2/concepts/arithmetization.html) it says:
> Typically, a configuration will define polynomial constraints that are switched off and on by selectors defined in fixed columns. For example, a constraint q<sub>i</sub>⋅p(...) = 0 can be switched off for a particular row i by setting q<sub>i</sub> = 0. In this case we sometimes refer to a set of constraints controlled by a set of selector columns that are designed to be used together, as a ***gate***.

That definition of a gate is selector-centric and covers all polynomial constraints that are enabled by a set of selectors. Even if we simplified it to the constraints covered by a single selector, I think it wouldn't be the same concept as what is being called a `Gate` in the code here, which only refers to a single polynomial constraint.

Perhaps just rename `Gate` to `PolyConstraint`?

Reply:
That interpretation doesn't match the next sentence:

> Typically there will be a standard gate that supports generic operations like field multiplication and division

The standard PLONK gate is a single polynomial constraint containing a set of sub-constraints controlled by a set of selector columns. We currently refer to that as a `Gate`, and it is what we can create with `meta.create_gate`. That is a different interpretation than "a gate is the set of all polynomial constraints involving the same selector", which is what I think you are thinking of.

Reply:
The standard PLONK gate does satisfy that definition. Its selectors are `sa`, `sb`, `sc`, and `sm`, and it has a (single-element) set of polynomial constraints controlled by those selectors.

Reply:
Addressed this in #273.